### PR TITLE
[RFR] Introduce BoardState objects and pop env method

### PR DESCRIPTION
**Pull request by Anko59**
Test with:
```
from env import Board
board = Board(['Aude', 'Steven', 'Kevin', 'Rachida'])
print('Board start state: notice, next_player cards are shown')
print(board.get_state())
print('Board legal_moves:')
print(board.legal_moves)
print('Aude plays flower')
board.push('F')
print('Her state now: ')
print(board.get_state().players[0])
print('Steven plays flower')
board.push('F')
print('Kevin plays skull')
board.push('S')
print('Rachida plays flower')
board.push('F')
print("New state at the end of first round: ")
print(board.get_state())
print('Player can now bet, legal moves are')
print(board.legal_moves)
print('Aude adds one card, Steven bets 2')
board.push('F')
board.push('B2')
print('Bet holder is Steven, highest_bet is 2, only bet actions available')
print(board.get_state().bet_holder)
print(board.get_state().highest_bet)
print(board.legal_moves)
print('Kevin bets 3, every other passes, his card is shown He skulled himself')
board.push('B3')
board.push('P')
board.push('P')
board.push('P')
print(board.get_state())
print(board.legal_moves)
print('He loses a flower, See how he is the next player: ')
board.push('LF')
print(board.get_state().next_player)
print('next Round, every one plays flower, Kevin fulls and can show cards')
board.push('F')
board.push('F')
board.push('F')
board.push('F')
board.push('B4')
board.push('P')
board.push('P')
board.push('P')
print(board.get_state())
print(board.legal_moves)
print('He shows all the cards in an order he chooss')
board.push('RAude')
board.push('P')
board.push('P')
board.push('P')
board.push('RSteven')
board.push('P')
board.push('P')
board.push('P')
board.push('RRachida')
print('He now has one point')
print(board.get_state().players[0].points)
print('Actually regret')
board.pop()
print(board.get_state().players[0].points)
```

### CODE REVIEW

File: env/actions.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+from .card import Card
+from dataclasses import dataclass
+from abc import ABC, abstractmethod
+
+
+class Action(ABC):
+
+    @property
+    @abstractmethod
+    def notation(self) -> str:
```
Comment (Smirkey) [inline env/actions.py]: Maybe we could use magic methods here

```suggestion
    def __str__(self) -> str:
```

File: env/actions.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+class PassAction(Action):
+    pass
+    notation: str = "P"
+
+
+class NotationDoesNotExistException(Exception):
+    pass
+
+
+def parse_notation(notation: str) -> Action:
```
Comment (Smirkey) [inline env/actions.py]: Noice

File: env/board.py [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+from .actions import (
+    Action,
+    PlayCardAction,
+    BetAction,
+    RevealCardAction,
+    LoseCardAction,
+    PassAction,
+    parse_notation,
+)
+from typing import List, Optional, Union
```
Comment (Smirkey) [inline env/board.py:11]: If we use python 3.10 or above we can use the builtin ˋlist` for typing 
And we can also replace Union with 
ObjA | ObjB

File: env/board.py [hunk 4]
```diff
@@ -114,25 +106,60 @@ def _process_action(self, player: Player, action: Action):
...
                 (self.players.index(self.next_player) + 1) % len(self.players)
             ]
-        self.legal_moves = self._get_legal_moves(self.next_player)
-        if not self._are_more_than_two_players_alive():
-            self.winner = [x for x in self.players if x.alive][0]  # type: ignore
-        elif self._has_anyone_won():
-            self.winner = [x for x in self.players if x.points > 1][0]  # type: ignore
+        self.legal_moves = self._get_legal_moves()
+
+    def pop(self):
```
Comment (Smirkey) [inline env/board.py:155]: Trop bien

File: env/card.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+class Card:
```
Comment (Smirkey) [inline env/card.py]: Cards could be an enum

File: env/player.py [hunk 6]
```diff
@@ -15,10 +17,23 @@ class NoFlowerException(Exception):
...
-    hidden = "HIDDEN"
+class CannotLoadHiddenStateException(Exception):
+    pass
+
+
+@dataclass
+class PlayerState:
+    name: str
+    cards_hand: List[str]
+    cards_stack: List[str]
```
Comment (Smirkey) [inline env/player.py]: 🤔
Would be nice if those objects were List[Card]

**Review by Smirkey — APPROVED** [review]
Only minor comments
👏

File: env/player.py (hunk 6 above)
Comment (Anko59) [reply env/player.py]: Use to be the case but mypy was not happy

**Pull request merged**